File: rbp-algebra-python/bridges.py
"""
RBP Bridges — The 6-Stage SPEC-BRIDGE Waterway
================================================
Defines the 6 BRIDGE gates that form the Reflect layer.

Architecture:
  EB_VECTOR (source)
    → SPEC_LINE (per-pesticide vertical pipes)
    → L1 SPEC-BRIDGE-TARGET     target mismatch → FullBlock
    → L2 SPEC-BRIDGE-USAGE      usage limit     → FullBlock
    → L3 SPEC-BRIDGE-PHI        PHI insufficient → Attenuate 0.5
    → L4 SPEC-BRIDGE-ROTATION   rotation abuse   → Attenuate 0.3
    → L5 SPEC-BRIDGE-MIXING     mixing banned    → FullBlock
    → L6 SPEC-BRIDGE-TOXICITY   highly toxic     → Attenuate 0.7
    → flowing lines → set enumeration → SPEC_BOX
"""

from __future__ import annotations
import sys, os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from rbp_types import (
    AttenuateValue, Bridge, BridgeContext, Pesticide, ToxicityClass,
    WeightAction,
)


def _full_pass(_ctx: BridgeContext) -> WeightAction:
    return WeightAction.FULL_PASS


def _full_block(_ctx: BridgeContext) -> WeightAction:
    return WeightAction.FULL_BLOCK


def _attenuate(factor: float):
    """Return a weight_fn that produces AttenuateValue(factor)."""
    def fn(_ctx: BridgeContext) -> AttenuateValue:
        return AttenuateValue(factor)
    return fn


# =============================================================================
# The 6 SPEC-BRIDGES
# =============================================================================

spec_bridges: list[Bridge] = [
    # L1: Target matching
    Bridge(
        bridge_id="SPEC-BRIDGE-TARGET",
        level=1.0,
        description="Blocks pesticides whose target diseases don't overlap with entry vector",
        weight_fn=lambda ctx: (
            WeightAction.FULL_PASS if ctx.target_match > 0
            else WeightAction.FULL_BLOCK
        ),
        reason_fn=lambda ctx: f"{ctx.pesticide.name}: target disease does not match entry vector",
        warning_fn=lambda _: "",
    ),

    # L2: Usage limit
    Bridge(
        bridge_id="SPEC-BRIDGE-USAGE",
        level=2.0,
        description="Blocks pesticides that have reached their annual application limit",
        weight_fn=lambda ctx: (
            WeightAction.FULL_BLOCK
            if ctx.pesticide.max_applications != -1
            and ctx.usage_state.get(ctx.pesticide.pid, 0) >= ctx.pesticide.max_applications
            else WeightAction.FULL_PASS
        ),
        reason_fn=lambda ctx: (
            f"Application limit reached "
            f"({ctx.usage_state.get(ctx.pesticide.pid, 0)}/{ctx.pesticide.max_applications if ctx.pesticide.max_applications != -1 else 'unlimited'})"
        ),
        warning_fn=lambda _: "",
    ),

    # L3: PHI (Pre-Harvest Interval)
    Bridge(
        bridge_id="SPEC-BRIDGE-PHI",
        level=3.0,
        description="Attenuates pesticides where PHI residual days are insufficient",
        weight_fn=lambda ctx: (
            _attenuate(0.5)(ctx)
            if ctx.interval_days is not None and ctx.interval_days < ctx.pesticide.phi_days
            else WeightAction.FULL_PASS
        ),
        reason_fn=lambda _: "PHI check: not a blocker, only attenuator",
        warning_fn=lambda ctx: (
            f"{ctx.pesticide.name}: PHI residual days check required "
            f"(last spray {ctx.interval_days} days ago, PHI {ctx.pesticide.phi_days} days)"
            if ctx.interval_days is not None else ""
        ),
        penalty=("safety", -10.0),
    ),

    # L4: Rotation management
    Bridge(
        bridge_id="SPEC-BRIDGE-ROTATION",
        level=4.0,
        description="Attenuates pesticides in systems with excessive consecutive use",
        weight_fn=lambda ctx: (
            _attenuate(0.3)(ctx)
            if ctx.pesticide.system_code not in ("MIX", "PHYSICAL")
            and ctx.rotation_state.get(ctx.pesticide.system_code, 0) >= 2
            else WeightAction.FULL_PASS
        ),
        reason_fn=lambda ctx: (
            f"{ctx.pesticide.name}: same system ({ctx.pesticide.system_code}) "
            f"used {ctx.rotation_state.get(ctx.pesticide.system_code, 0)} times consecutively"
        ),
        warning_fn=lambda ctx: (
            f"{ctx.pesticide.name}: same system ({ctx.pesticide.system_name}) "
            f"used {ctx.rotation_state.get(ctx.pesticide.system_code, 0)} times consecutively (resistance risk)"
        ),
        penalty=("resistance", -15.0),
    ),

    # L5: Mixing compatibility
    Bridge(
        bridge_id="SPEC-BRIDGE-MIXING",
        level=5.0,
        description="Blocks pesticides that conflict with the last sprayed pesticide",
        weight_fn=lambda ctx: (
            WeightAction.FULL_BLOCK
            if any(_has_mixing_conflict(ctx.pesticide, lp) for lp in ctx.last_pesticides)
            else WeightAction.FULL_PASS
        ),
        reason_fn=lambda ctx: (
            f"{ctx.pesticide.name} cannot mix with last sprayed pesticides"
        ),
        warning_fn=lambda _: "",
    ),

    # L6: Toxicity class
    Bridge(
        bridge_id="SPEC-BRIDGE-TOXICITY",
        level=6.0,
        description="Attenuates highly toxic pesticides (discouraged but not prohibited)",
        weight_fn=lambda ctx: (
            _attenuate(0.7)(ctx)
            if ctx.pesticide.toxicity_class == ToxicityClass.HIGHLY_TOXIC
            else WeightAction.FULL_PASS
        ),
        reason_fn=lambda _: "Toxicity check: not a blocker, only attenuator",
        warning_fn=lambda ctx: f"{ctx.pesticide.name}: highly toxic classification",
        penalty=("safety", -8.0),
    ),
]


def _has_mixing_conflict(a: Pesticide, b: Pesticide) -> bool:
    """Check if two pesticides have a mixing conflict."""
    def _mentions(haystack: str, needle: str) -> bool:
        return needle in haystack or needle.lower() in haystack.lower()

    a_bans = a.mixing_ban_targets
    b_bans = b.mixing_ban_targets

    # Does A ban mixing with B's system or name?
    if any(_mentions(b.system_name, t) or _mentions(b.name, t) for t in a_bans):
        return True
    # Does B ban mixing with A's system or name?
    if any(_mentions(a.system_name, t) or _mentions(a.name, t) for t in b_bans):
        return True
    return False


def set_has_internal_mixing_conflict(pesticides: list[Pesticide]) -> bool:
    """Check if a set of pesticides has internal mixing conflicts."""
    if len(pesticides) != 2:
        return False
    return _has_mixing_conflict(pesticides[0], pesticides[1])


def run_spec_line(
    pesticide: Pesticide,
    entry_vector,  # EntryVector
    ctx: BridgeContext,
) -> FlowResult:
    """Run a single pesticide's SPEC_LINE through all 6 bridges."""
    ctx_with_p = BridgeContext(
        pesticide=pesticide,
        entry_vector=entry_vector,
        target_match=ctx.target_match,
        usage_state=ctx.usage_state,
        last_spray_date=ctx.last_spray_date,
        last_pesticide_ids=ctx.last_pesticide_ids,
        last_pesticides=ctx.last_pesticides,
        interval_days=ctx.interval_days,
        rotation_state=ctx.rotation_state,
    )
    return run_line_through_bridges(entry_vector, spec_bridges, ctx_with_p)

File: rbp-algebra/src/Data/RBP/Bridges.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE RecordWildCards #-}
{-|
Module      : Data.RBP.Bridges
Description : The 6-stage SPEC-BRIDGE waterway definition.

This module defines the 6 BRIDGE gates that form the Reflect layer.
Each gate is a pure function from BridgeContext → WeightAction.

The algebraic structure:

  EB_VECTOR (source)
    → SPEC_LINE (per-pesticide vertical pipes)
    → L1 SPEC-BRIDGE-TARGET     target mismatch → FullBlock
    → L2 SPEC-BRIDGE-USAGE      usage limit     → FullBlock
    → L3 SPEC-BRIDGE-PHI        PHI insufficient → Attenuate 0.5
    → L4 SPEC-BRIDGE-ROTATION   rotation abuse   → Attenuate 0.3
    → L5 SPEC-BRIDGE-MIXING     mixing banned    → FullBlock
    → L6 SPEC-BRIDGE-TOXICITY   highly toxic     → Attenuate 0.7
    → flowing lines → set enumeration → SPEC-BRIDGE-MIXING-SET → SPEC_BOX

Each bridge is defined declaratively. The \"business logic\" (which
conditions trigger blocking vs attenuation) lives in the weightFn
as a pattern match on BridgeContext — NOT in the engine loop.

This is the key proof point: procedural if/else chains become
declarative algebraic transformations.
-}
module Data.RBP.Bridges
  ( specBridges
  , runSpecLine
  , runAllSpecLines
  , hasMixingConflict
  , setHasInternalMixingConflict
  , buildMixingReason
  ) where

import Data.RBP.Types
import Data.RBP.Core (runLineThroughBridges)
import qualified Data.Map.Strict as Map
import Data.List (sortBy)
import Data.Ord (comparing)
import Data.Char (toLower)

------------------------------------------------------------------------------
-- Bridge helper constructors
------------------------------------------------------------------------------

fullPass :: Int -> WeightAction
fullPass _ = FullPass

fullBlock :: Int -> WeightAction
fullBlock _ = FullBlock

attenuate :: Double -> Int -> WeightAction
attenuate w _ = Attenuate w

------------------------------------------------------------------------------
-- The 6 SPEC-BRIDGES
------------------------------------------------------------------------------

{-|
All 6 SPEC-BRIDGES in level order.

Each bridge is a record with:
  - id, level, direction (structural invariants)
  - weightFn: the business logic as a pure function
  - reasonFn: human-readable explanation when blocked
  - penalty: (axis, delta) for scoring when attenuated
  - warningFn: human-readable warning when attenuated
  - description: documentation

The weightFn is where the \"if/else\" of the original code lives —
but as a pattern match on BridgeContext, not as control flow in the engine.
-}
specBridges :: [Bridge]
specBridges =
  [ -- L1: Target matching — does this pesticide target any active disease?
    Bridge
      { bid = BID "SPEC-BRIDGE-TARGET"
      , bLevel = 1.0
      , bDirection = ForwardOnly
      , bWeightFn = \ctx ->
          if targetMatchCtx ctx > 0
            then fullPass (vectorDim (bcEntryVector ctx))
            else fullBlock (vectorDim (bcEntryVector ctx))
      , bReasonFn = \ctx ->
          pname (bcPesticide ctx) ++ ": target disease does not match entry vector"
      , bPenalty = Nothing
      , bWarningFn = \_ -> ""
      , bDescription = "Blocks pesticides whose target diseases don't overlap with entry vector"
      }

  , -- L2: Usage limit — annual application count exceeded?
    Bridge
      { bid = BID "SPEC-BRIDGE-USAGE"
      , bLevel = 2.0
      , bDirection = ForwardOnly
      , bWeightFn = \ctx ->
          let usageCount = Map.findWithDefault 0 (pid (bcPesticide ctx)) (bcUsageState ctx)
              lim = maxApplications (bcPesticide ctx)
              limitReached = lim /= -1 && usageCount >= lim  -- -1 = Infinity
          in if limitReached
               then fullBlock (vectorDim (bcEntryVector ctx))
               else fullPass (vectorDim (bcEntryVector ctx))
      , bReasonFn = \ctx ->
          let usageCount = Map.findWithDefault 0 (pid (bcPesticide ctx)) (bcUsageState ctx)
              lim = maxApplications (bcPesticide ctx)
          in "Application limit reached (" ++ show usageCount ++ "/" ++
             if lim == -1 then "unlimited" else show lim ++ ")"
      , bPenalty = Nothing
      , bWarningFn = \_ -> ""
      , bDescription = "Blocks pesticides that have reached their annual application limit"
      }

  , -- L3: PHI (Pre-Harvest Interval) — harvest too soon after last spray?
    Bridge
      { bid = BID "SPEC-BRIDGE-PHI"
      , bLevel = 3.0
      , bDirection = ForwardOnly
      , bWeightFn = \ctx ->
          case (bcIntervalDays ctx, phiDays (bcPesticide ctx)) of
            (Just days, phi) ->
              if days < phi
                then attenuate 0.5 (vectorDim (bcEntryVector ctx))
                else fullPass (vectorDim (bcEntryVector ctx))
            (Nothing, _) ->
              fullPass (vectorDim (bcEntryVector ctx))  -- no history = safe
      , bReasonFn = \_ -> "PHI check: not a blocker, only attenuator"
      , bPenalty = Just ("safety", -10.0)
      , bWarningFn = \ctx ->
          case bcIntervalDays ctx of
            Just days -> pname (bcPesticide ctx) ++ ": PHI residual days check required (" ++
                         show days ++ " days since last spray, PHI " ++
                         show (phiDays (bcPesticide ctx)) ++ " days)"
            Nothing -> ""
      , bDescription = "Attenuates pesticides where PHI residual days are insufficient (not a hard block)"
      }

  , -- L4: Rotation management — same system used 2+ times consecutively?
    Bridge
      { bid = BID "SPEC-BRIDGE-ROTATION"
      , bLevel = 4.0
      , bDirection = ForwardOnly
      , bWeightFn = \ctx ->
          let sysCode = systemCode (bcPesticide ctx)
              isNonRotation = sysCode `elem` ["MIX", "PHYSICAL"]
              rotCount = Map.findWithDefault 0 sysCode (bcRotationState ctx)
              abuse = rotCount >= 2 && not isNonRotation
          in if abuse
               then attenuate 0.3 (vectorDim (bcEntryVector ctx))
               else fullPass (vectorDim (bcEntryVector ctx))
      , bReasonFn = \ctx ->
          let sysCode = systemCode (bcPesticide ctx)
              rotCount = Map.findWithDefault 0 sysCode (bcRotationState ctx)
          in pname (bcPesticide ctx) ++ ": same system (" ++ sysCode ++ ") used " ++
             show rotCount ++ " times consecutively (resistance risk)"
      , bPenalty = Just ("resistance", -15.0)
      , bWarningFn = \ctx ->
          let sysCode = systemCode (bcPesticide ctx)
              rotCount = Map.findWithDefault 0 sysCode (bcRotationState ctx)
          in pname (bcPesticide ctx) ++ ": same system (" ++ system (bcPesticide ctx) ++
             ") used " ++ show rotCount ++ " times consecutively (resistance risk)"
      , bDescription = "Attenuates pesticides in systems with excessive consecutive use (resistance management)"
      }

  , -- L5: Mixing compatibility — conflicts with last sprayed pesticide?
    Bridge
      { bid = BID "SPEC-BRIDGE-MIXING"
      , bLevel = 5.0
      , bDirection = ForwardOnly
      , bWeightFn = \ctx ->
          let lastPests = bcLastPesticides ctx
              conflicts = any (hasMixingConflict (bcPesticide ctx)) lastPests
          in if conflicts
               then fullBlock (vectorDim (bcEntryVector ctx))
               else fullPass (vectorDim (bcEntryVector ctx))
      , bReasonFn = \ctx ->
          pname (bcPesticide ctx) ++ " cannot mix with last sprayed pesticides"
      , bPenalty = Nothing
      , bWarningFn = \_ -> ""
      , bDescription = "Blocks pesticides that conflict with the last sprayed pesticide"
      }

  , -- L6: Toxicity class — highly toxic (劇物)?
    Bridge
      { bid = BID "SPEC-BRIDGE-TOXICITY"
      , bLevel = 6.0
      , bDirection = ForwardOnly
      , bWeightFn = \ctx ->
          let isHighlyToxic = toxicityClass (bcPesticide ctx) == HighlyToxic
          in if isHighlyToxic
               then attenuate 0.7 (vectorDim (bcEntryVector ctx))
               else fullPass (vectorDim (bcEntryVector ctx))
      , bReasonFn = \_ -> "Toxicity check: not a blocker, only attenuator"
      , bPenalty = Just ("safety", -8.0)
      , bWarningFn = \ctx ->
          pname (bcPesticide ctx) ++ ": highly toxic classification"
      , bDescription = "Attenuates highly toxic pesticides (discouraged but not prohibited)"
      }
  ]

------------------------------------------------------------------------------
-- Running specs
------------------------------------------------------------------------------

{-|
Run a single pesticide's SPEC_LINE through all 6 bridges.
-}
runSpecLine
  :: Pesticide
  -> EntryVector
  -> BridgeContext
  -> FlowResult
runSpecLine pesticide ebVector ctx =
  let ctx' = ctx { bcPesticide = pesticide }
  in runLineThroughBridges ebVector specBridges ctx'

{-|
Run all pesticides through their SPEC_LINEs.
-}
runAllSpecLines
  :: [Pesticide]
  -> EntryVector
  -> BridgeContext
  -> [FlowResult]
runAllSpecLines pesticides ebVector ctx =
  map (\p -> runSpecLine p ebVector ctx { bcPesticide = p }) pesticides

------------------------------------------------------------------------------
-- Mixing conflict helpers
------------------------------------------------------------------------------

{-|
Check if two pesticides have a mixing conflict.
Two pesticides conflict if either bans mixing with the other's system or name.
-}
hasMixingConflict :: Pesticide -> Pesticide -> Bool
hasMixingConflict a b =
  let aBans = mixingBanTargets a
      bBans = mixingBanTargets b
      bSys = system b
      bNm = pname b
      aSys = system a
      aNm = pname a
  in (any (\t -> bSys == t || t `isInfixOf` bSys || bNm == t || t `isInfixOf` bNm) aBans) ||
     (any (\t -> aSys == t || t `isInfixOf` aSys || aNm == t || t `isInfixOf` aNm) bBans)

{-|
Check if a set of pesticides has internal mixing conflicts.
Only applies to pairs (2-element sets).
-}
setHasInternalMixingConflict :: [Pesticide] -> Bool
setHasInternalMixingConflict [a, b] = hasMixingConflict a b
setHasInternalMixingConflict _      = False

{-|
Build human-readable mixing-conflict reason strings between two pesticides.
Mirrors the Python reference engine's main.py:_build_mixing_reason /
_mentions exactly (haystack = ban-target string, needle = name/system) —
note this argument order is intentionally the reverse of 'hasMixingConflict'
above, matching a pre-existing asymmetry in the Python reference itself.
-}
buildMixingReason :: Pesticide -> Pesticide -> [String]
buildMixingReason a b =
  concat
    [ [ pname a ++ "は" ++ pname b ++ "（" ++ system b ++ "）と混用不可"
      | any (\t -> mentions t (system b) || mentions t (pname b)) (mixingBanTargets a) ]
    , [ pname b ++ "は" ++ pname a ++ "（" ++ system a ++ "）と混用不可"
      | any (\t -> mentions t (system a) || mentions t (pname a)) (mixingBanTargets b) ]
    ]
  where
    mentions haystack needle =
      needle `isInfixOf` haystack || map toLower needle `isInfixOf` map toLower haystack

------------------------------------------------------------------------------
-- Helpers
------------------------------------------------------------------------------

lookupPid :: PesticideId -> Pesticide
lookupPid _ = error "lookupPid: not implemented — mixing check uses direct Pesticide comparison"

targetMatchCtx :: BridgeContext -> Int
targetMatchCtx (BridgeContext { bcTargetMatch = TM m }) = m

isInfixOf :: String -> String -> Bool
isInfixOf needle haystack = any (isPrefixOf needle) (tails haystack)

isPrefixOf :: String -> String -> Bool
isPrefixOf [] _ = True
isPrefixOf _ [] = False
isPrefixOf (x:xs) (y:ys) = x == y && isPrefixOf xs ys

tails :: String -> [String]
tails [] = [""]
tails s@( _:rest) = s : tails rest
